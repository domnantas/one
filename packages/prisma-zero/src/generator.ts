import type { DMMF } from '@prisma/generator-helper'
import { generatorHandler, type GeneratorOptions } from '@prisma/generator-helper'
import { writeFile, mkdir, readFile } from 'node:fs/promises'
import { join } from 'node:path'
import { createHash } from 'node:crypto'
// @ts-ignore
import { version } from '../package.json'

type Config = {
  name: string
  prettier: boolean
  resolvePrettierConfig: boolean
  schemaVersion?: number
}

function mapPrismaTypeToZeroType(field: DMMF.Field): string {
  const typeMap: Record<string, string> = {
    String: 'string',
    Boolean: 'boolean',
    Int: 'number',
    Float: 'number',
    DateTime: 'number', // Zero uses timestamps
    Json: 'json',
    BigInt: 'number',
    Decimal: 'number',
  }

  let output = `${field.name}: `
  if (field.kind === 'enum') {
    output += `enumeration<${field.type}>(${field.isRequired ? '' : 'true'})`
    return output
  }

  // If it's not an enum, we need to handle required and not required fields differently
  const value = typeMap[field.type] || 'string'
  if (field.isRequired) {
    output += `"${value}"`
  } else {
    output += `{ type: "${value}", optional: true }`
  }

  return output
}

function generateRelationships(model: DMMF.Model, dmmf: DMMF.Document) {
  const relationships: string[] = []

  model.fields
    .filter((field) => field.relationName)
    .forEach((field) => {
      const relName = field.name
      let sourceField: string
      let destField: string

      if (field.isList) {
        // For "many" side relationships, we need to find the matching field in the target model
        // that references back to this model
        const targetModel = dmmf.datamodel.models.find((m) => m.name === field.type)
        const backReference = targetModel?.fields.find(
          (f) => f.relationName === field.relationName && f.type === model.name
        )
        sourceField = 'id'
        destField = backReference?.relationFromFields?.[0] || 'id'
      } else {
        // For "one" side relationships, use the foreign key
        sourceField = field.relationFromFields?.[0] || 'id'
        destField = field.relationToFields?.[0] || 'id'
      }

      const destModel = field.type

      relationships.push(`    ${relName}: {
      sourceField: "${sourceField}",
      destField: "${destField}",
      destSchema: () => ${destModel}Schema,
    }`)
    })

  return relationships.join(',\n')
}

function getTableName(model: DMMF.Model) {
  return model.dbName || model.name
}

function generateSchemaHash(models: DMMF.Model[], enums: DMMF.DatamodelEnum[]): string {
  const hash = createHash('sha256')

  // Only hash the structural elements that affect the schema
  const schemaStructure = {
    models: models.map((model) => ({
      name: model.name,
      dbName: model.dbName,
      fields: model.fields.map((f) => ({
        // Only include field properties that affect the schema
        name: f.name,
        type: f.type,
        isRequired: f.isRequired,
        isList: f.isList,
        relationName: f.relationName,
        relationFromFields: f.relationFromFields,
        relationToFields: f.relationToFields,
        default: f.default,
        unique: f.isUnique,
      })),
      primaryKey: model.primaryKey,
      uniqueFields: model.uniqueFields,
      uniqueIndexes: model.uniqueIndexes,
    })),
    enums: enums.map((enumType) => ({
      name: enumType.name,
      values: enumType.values.map((v) => ({
        name: v.name,
        dbName: v.dbName,
      })),
    })),
  }

  hash.update(JSON.stringify(schemaStructure))
  return hash.digest('hex')
}

async function getCurrentVersion(
  outputDir: string,
  filename: string
): Promise<{ version: number; hash: string | null }> {
  try {
    const content = await readFile(join(outputDir, filename), 'utf-8')
    const versionMatch = content.match(/version:\s*(\d+)/)
    const hashMatch = content.match(/Schema hash: ([a-f0-9]+)/)

    return {
      version: versionMatch ? Number.parseInt(versionMatch[1], 10) : 0,
      hash: hashMatch ? hashMatch[1] : null,
    }
  } catch {
    return { version: 0, hash: null }
  }
}

// Export the onGenerate function separately
export async function onGenerate(options: GeneratorOptions) {
  const { generator, dmmf } = options

  const outputFile = 'schema.ts'
  const outputDir = generator.output?.value

  if (!outputDir) {
    throw new Error('Output directory is required')
  }

  // Generate hash and get current version
  const newHash = generateSchemaHash([...dmmf.datamodel.models], [...dmmf.datamodel.enums])
  const { version: currentVersion, hash: currentHash } = await getCurrentVersion(
    outputDir,
    outputFile
  )
  const nextAutoincrementVersion = currentHash !== newHash ? currentVersion + 1 : currentVersion

  if (generator.config.schemaVersion && Number.isNaN(Number(generator.config.schemaVersion))) {
    throw new Error('Schema version must be a number')
  }

  const config = {
    name: generator.name,
    prettier: generator.config.prettier === 'true', // Default false,
    resolvePrettierConfig: generator.config.resolvePrettierConfig !== 'false', // Default true
    schemaVersion: generator.config.schemaVersion
      ? Number(generator.config.schemaVersion)
      : nextAutoincrementVersion,
  } satisfies Config

  const enums = dmmf.datamodel.enums
  const models = dmmf.datamodel.models.filter((model) => {
    const tags = extractTags(model.documentation)
    if (tags.includes('private')) {
      return false
    }
    return true
  })

  let output = `// Generated by @vxrn/prisma-zero\n\n`

  output += "import { createSchema, type Row, column } from '@rocicorp/zero'\n\n"
  output += 'const { enumeration } = column;\n\n'

  if (enums.length > 0) {
    output += '// Define enums\n\n'
    // Generate enums
    enums.forEach((enumType) => {
      // Generate TypeScript enum
      output += `export enum ${enumType.name} {\n`
      enumType.values.forEach((value) => {
        // Handle mapped values
        const enumValue = value.dbName || value.name
        output += `  ${value.name} = "${enumValue}",\n`
      })
      output += '}\n\n'
    })
  }

  if (models.length > 0) {
    output += '// Define schemas\n\n'

    // Generate schemas for models
    models.forEach((model) => {
      output += `const ${model.name}Schema = {\n`
      output += `  tableName: "${getTableName(model)}",\n`
      output += '  columns: {\n'

      model.fields
        .filter((field) => !field.relationName) // Skip relation fields
        .forEach((field) => {
          const fieldValue = mapPrismaTypeToZeroType(field)
          output += `    ${fieldValue},\n`
        })

      output += '  },\n'

      // Add relationships if any exist
      const relationships = generateRelationships(model, dmmf)
      if (relationships) {
        output += '  relationships: {\n'
        output += relationships
        output += '\n  },\n'
      }

      // Add primary key
      const primaryKey = model.primaryKey?.fields
        ? model.primaryKey.fields
        : model.fields.find((f) => f.isId)?.name

      if (!primaryKey) {
        throw new Error(`No primary key found for ${model.name}`)
      }

      const primaryKeyString = JSON.stringify(primaryKey)

      output += `  primaryKey: ${primaryKeyString},\n`
      output += '} as const;\n\n'
    })
  }

  output += '// Define schema\n\n'
  // Generate the main schema export
  output += 'export const schema = createSchema({\n'
  output += `  version: ${config.schemaVersion},\n`
  output += '  tables: {\n'
  models.forEach((model) => {
    output += `    ${getTableName(model)}: ${model.name}Schema,\n`
  })
  output += '  },\n'
  output += '});\n\n'

  // Generate types
  output += '\n// Define types\n'
  output += 'export type Schema = typeof schema;\n'
  models.forEach((model) => {
    output += `export type ${model.name} = Row<typeof ${model.name}Schema>;\n`
  })

  output += '\n// Schema hash: ' + newHash + '\n'

  // Ensure output directory exists
  if (outputDir) {
    await mkdir(outputDir, { recursive: true })
  }

  // Write the output to a file
  if (outputDir) {
    await writeFile(join(outputDir, outputFile), output)
  }
}

// Use the exported function in the generator handler
generatorHandler({
  onManifest() {
    return {
      version,
      defaultOutput: 'generated/zero',
      prettyName: 'Zero Schema',
    }
  },
  onGenerate,
})

function extractTags(doc?: string): string[] {
  if (!doc) return []
  const match = doc.match(/@tag\(([^)]+)\)/)
  if (!match) return []
  return match[1].split(',').map((tag) => tag.trim().replace(/['"]/g, ''))
}
