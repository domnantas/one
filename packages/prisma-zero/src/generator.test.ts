import { describe, it, expect, beforeEach, vi } from 'vitest'
import type { DMMF, GeneratorOptions } from '@prisma/generator-helper'
import * as fs from 'node:fs/promises'
import { onGenerate } from './generator'

// Mock fs/promises
vi.mock('node:fs/promises', () => ({
  writeFile: vi.fn(),
  mkdir: vi.fn(),
  readFile: vi.fn(),
}))

// Helper Functions
function createField(
  name: string,
  type: string,
  isRequired = true,
  isList = false,
  relationName?: string
): DMMF.Field {
  return {
    name,
    kind: 'scalar',
    type,
    isRequired,
    isList,
    relationName,
    isId: false,
    isReadOnly: false,
    isGenerated: false,
    isUpdatedAt: false,
    isUnique: false,
    hasDefaultValue: false,
    documentation: undefined,
    relationFromFields: [],
    relationToFields: [],
  }
}

function createTestOptions(dmmf: DMMF.Document): GeneratorOptions {
  return {
    generator: {
      output: { value: 'generated', fromEnvVar: null },
      name: 'test-generator',
      config: {},
      provider: { value: 'test-provider', fromEnvVar: null },
      binaryTargets: [],
      previewFeatures: [],
      sourceFilePath: '',
    },
    dmmf,
    schemaPath: '',
    datasources: [],
    otherGenerators: [],
    version: '0.0.0',
    datamodel: '',
  }
}

function createMockDMMF(models: DMMF.Model[], enums: DMMF.DatamodelEnum[] = []): DMMF.Document {
  return {
    datamodel: {
      models,
      enums,
      types: [],
      indexes: [],
    },
    schema: {} as any,
    mappings: {} as any,
  }
}

function createModel(
  name: string,
  fields: DMMF.Field[],
  options: Partial<DMMF.Model> = {}
): DMMF.Model {
  return {
    name,
    dbName: null,
    fields,
    uniqueFields: [],
    uniqueIndexes: [],
    primaryKey: null,
    ...options,
  }
}

describe('Generator', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe('Schema Generation', () => {
    it('should generate correct schema for basic model', async () => {
      const mockModel: DMMF.Model = {
        name: 'User',
        dbName: null,
        fields: [
          { ...createField('id', 'String'), isId: true },
          createField('name', 'String'),
          createField('email', 'String'),
          { ...createField('age', 'Int'), isRequired: false },
        ],
        uniqueFields: [],
        uniqueIndexes: [],
        primaryKey: null,
      }

      await onGenerate(createTestOptions(createMockDMMF([mockModel])))

      // Verify mkdir was called
      expect(fs.mkdir).toHaveBeenCalledWith('generated', { recursive: true })

      // Verify writeFile was called with correct schema
      const writeFileCalls = vi.mocked(fs.writeFile).mock.calls
      expect(writeFileCalls.length).toBe(1)

      const [, contentBuffer] = writeFileCalls[0]
      const content = contentBuffer.toString()

      expect(content).toMatchSnapshot()
    })

    it('should handle enums correctly', async () => {
      const mockEnum: DMMF.DatamodelEnum = {
        name: 'Role',
        values: [
          { name: 'USER', dbName: null },
          { name: 'ADMIN', dbName: null },
        ],
        dbName: null,
      }

      const mockModel: DMMF.Model = {
        name: 'User',
        dbName: null,
        fields: [
          { ...createField('id', 'String'), isId: true },
          { ...createField('role', 'Role'), kind: 'enum' },
        ],
        uniqueFields: [],
        uniqueIndexes: [],
        primaryKey: null,
      }

      await onGenerate(createTestOptions(createMockDMMF([mockModel], [mockEnum])))

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[0]
      const content = contentBuffer.toString()

      expect(content).toMatchSnapshot()
    })

    it('should handle relationships correctly', async () => {
      const userModel = createModel('User', [
        { ...createField('id', 'String'), isId: true },
        {
          ...createField('posts', 'Post', true, true),
          kind: 'object',
          relationName: 'UserPosts',
          relationToFields: ['id'],
          relationFromFields: ['userId'],
        },
      ])

      const postModel = createModel('Post', [
        { ...createField('id', 'String'), isId: true },
        {
          ...createField('userId', 'String'),
          kind: 'scalar',
          relationName: 'UserPosts',
          relationFromFields: ['userId'],
          relationToFields: ['id'],
        },
        {
          ...createField('user', 'User'),
          kind: 'object',
          relationName: 'UserPosts',
          relationFromFields: ['userId'],
          relationToFields: ['id'],
        },
      ])

      await onGenerate(createTestOptions(createMockDMMF([userModel, postModel])))

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[0]
      const content = contentBuffer.toString()

      expect(content).toMatchSnapshot()
    })
  })

  describe('Schema Version Management', () => {
    it('should increment version when schema hash changes', async () => {
      // First generation with simpler model creation
      const initialModel = createModel('User', [
        { ...createField('id', 'String'), isId: true },
        createField('name', 'String'),
      ])

      // Mock initial file read to simulate no existing file
      vi.mocked(fs.readFile).mockRejectedValueOnce(new Error('File not found'))

      await onGenerate(createTestOptions(createMockDMMF([initialModel])))

      // Get the first generated schema
      const [, firstContentBuffer] = vi.mocked(fs.writeFile).mock.calls[0]
      const firstContent = firstContentBuffer.toString()

      // Second generation with modified model
      const modifiedModel = createModel('User', [
        { ...createField('id', 'String'), isId: true },
        createField('name', 'String'),
        createField('email', 'String'), // Added field
      ])

      // Mock reading the generated file to be version 1
      vi.mocked(fs.readFile).mockResolvedValueOnce(firstContent)

      await onGenerate(createTestOptions(createMockDMMF([modifiedModel])))

      // Get the content of the last write
      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[1]
      const content = contentBuffer.toString()

      // Version should be incremented
      expect(content).toContain('version: 2')
    })

    it('should not increment version when schema hash remains the same', async () => {
      const model = createModel('User', [
        { ...createField('id', 'String'), isId: true },
        createField('name', 'String'),
      ])

      const dmmf = createMockDMMF([model])

      // Mock reading existing schema with version 1
      vi.mocked(fs.readFile).mockResolvedValueOnce(
        Buffer.from('version: 1\n// Schema hash: abc123')
      )

      await onGenerate(createTestOptions(dmmf))

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[
        vi.mocked(fs.writeFile).mock.calls.length - 1
      ]
      const content = contentBuffer.toString()

      // Version should remain the same
      expect(content).toContain('version: 1')
    })

    it('should use provided schemaVersion from generator config', async () => {
      const model = createModel('User', [
        { ...createField('id', 'String'), isId: true },
        createField('name', 'String'),
      ])

      const options = createTestOptions(createMockDMMF([model]))
      // Set the schema version in generator config
      options.generator.config.schemaVersion = '42'

      await onGenerate(options)

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[
        vi.mocked(fs.writeFile).mock.calls.length - 1
      ]
      const content = contentBuffer.toString()

      // Version should match the provided config value
      expect(content).toContain('version: 42')
    })
  })
})
