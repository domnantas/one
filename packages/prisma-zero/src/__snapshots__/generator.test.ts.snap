// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generator > Schema Generation > should generate correct schema for basic model 1`] = `
"// Generated by Zero Schema Generator

import { createSchema, definePermissions, Row, column } from "@rocicorp/zero";

const { enumeration } = column;

// Define schemas

const UserSchema = {
  tableName: "User",
  columns: {
    id: "string",
    name: "string",
    email: "string",
    age: { type: "number", optional: true },
  },
  primaryKey: "id",
} as const;

// Define schema

export const schema = createSchema({
  version: 1,
  tables: {
    User: UserSchema,
  },
});

// Define types
export type Schema = typeof schema;
export type User = Row<typeof UserSchema>;

// Define permissions

// Important: currently no permissions are generated so evey operation is allowed!


export const permissions = definePermissions(schema, () => ({}));

// DO NOT TOUCH THIS. The schema hash is used to determine if the schema has changed and correctly update the version.
// Schema hash: 39e9a6851c0a6d2a9e610b80c7b3c979c237c4f921f37571491153d83c2eca82
"
`;

exports[`Generator > Schema Generation > should handle enums correctly 1`] = `
"// Generated by Zero Schema Generator

import { createSchema, definePermissions, Row, column } from "@rocicorp/zero";

const { enumeration } = column;

// Define enums

export enum Role {
  USER = "USER",
  ADMIN = "ADMIN",
}

// Define schemas

const UserSchema = {
  tableName: "User",
  columns: {
    id: "string",
    role: enumeration<Role>(),
  },
  primaryKey: "id",
} as const;

// Define schema

export const schema = createSchema({
  version: 1,
  tables: {
    User: UserSchema,
  },
});

// Define types
export type Schema = typeof schema;
export type User = Row<typeof UserSchema>;

// Define permissions

// Important: currently no permissions are generated so evey operation is allowed!


export const permissions = definePermissions(schema, () => ({}));

// DO NOT TOUCH THIS. The schema hash is used to determine if the schema has changed and correctly update the version.
// Schema hash: 1769528a165c494f19172ce090f95266b4b3473d762f8c3187da6d187eb5d61d
"
`;

exports[`Generator > Schema Generation > should handle relationships correctly 1`] = `
"// Generated by Zero Schema Generator

import { createSchema, definePermissions, Row, column } from "@rocicorp/zero";

const { enumeration } = column;

// Define schemas

const UserSchema = {
  tableName: "User",
  columns: {
    id: "string",
  },
  relationships: {
    posts: {
      sourceField: "id",
      destField: "userId",
      destSchema: () => PostSchema,
    }
  },
  primaryKey: "id",
} as const;

const PostSchema = {
  tableName: "Post",
  columns: {
    id: "string",
  },
  relationships: {
    userId: {
      sourceField: "userId",
      destField: "id",
      destSchema: () => StringSchema,
    },
    user: {
      sourceField: "userId",
      destField: "id",
      destSchema: () => UserSchema,
    }
  },
  primaryKey: "id",
} as const;

// Define schema

export const schema = createSchema({
  version: 1,
  tables: {
    User: UserSchema,
    Post: PostSchema,
  },
});

// Define types
export type Schema = typeof schema;
export type User = Row<typeof UserSchema>;
export type Post = Row<typeof PostSchema>;

// Define permissions

// Important: currently no permissions are generated so evey operation is allowed!


export const permissions = definePermissions(schema, () => ({}));

// DO NOT TOUCH THIS. The schema hash is used to determine if the schema has changed and correctly update the version.
// Schema hash: 0e7d51a81b828c897fe0a887aa6bdc1d94e1706d5566ec51ee8e0bd9ce6099cc
"
`;
